pytorch_lightning/utilities/data.py

```diff
@@ -336,14 +353,29 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
             if param.name != "self" and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
         ]
 
-        cls = type(obj)
-        for arg_name, arg_value in chain(zip(param_names, args), kwargs.items()):
-            if hasattr(cls, arg_name) and getattr(cls, arg_name).fset is None:
-                # the class defines a read-only (no setter) property of this name. it's likely that the implementation
-                # will set `self._arg_name = arg_value` in `__init__` which is the attribute returned by the `arg_name`
-                # property so we are fine skipping in that case
-                continue
-            setattr(obj, arg_name, arg_value)
+        if not hasattr(obj, "_set_arg_names"):
+            set_arg_names = set()
+            for arg_name, arg_value in chain(zip(param_names, args), kwargs.items()):
+                if not hasattr(obj, "__" + arg_name):
+                    # Set the value privately if it has not been set yet. This achieves two things:
+                    # 1. The argument is the one that was passed in the outermost call
+                    #    and not overriden by a call to `super().__init__()`
+                    # 2. The argument is the passed value and does not get overwritten in `__init__()`
+                    setattr(obj, "__" + arg_name, arg_value)
+                    set_arg_names.add(arg_name)
+            obj._set_arg_names = set(set_arg_names)
+
+        if not hasattr(obj, "__dataset"):
+            # We have not found value for dataset yet, but we will find it eventually
+            # because it has to be passed to the original DataLoader
+            dataset_value = None
+            if "dataset" in param_names[: len(args)]:
+                dataset_value = args[param_names.index("dataset")]
+            elif "dataset" in kwargs:
+                dataset_value = kwargs["dataset"]
+            if dataset_value is not None:
+                setattr(obj, "__dataset", dataset_value)
```

Comment:
Why does this need to happen in this function?

It makes it coupled to dataloaders

Reply:
I separated it dataloader specifics into its own function